File: lib/services/cloud_account_service.dart
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;
import 'package:path/path.dart' as path;
import 'package:path_provider/path_provider.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../utils/api_config.dart';

class CloudSession {
  final int userId;
  final String username;
  final String token;

  const CloudSession({
    required this.userId,
    required this.username,
    required this.token,
  });
}

class CloudBackupInfo {
  final int id;
  final String fileName;
  final int payloadSize;
  final String uploadDate;
  final String createdAt;

  const CloudBackupInfo({
    required this.id,
    required this.fileName,
    required this.payloadSize,
    required this.uploadDate,
    required this.createdAt,
  });

  factory CloudBackupInfo.fromJson(Map<String, dynamic> json) {
    return CloudBackupInfo(
      id: json['id'] as int,
      fileName: (json['file_name'] ?? '').toString(),
      payloadSize: (json['payload_size'] as num?)?.toInt() ?? 0,
      uploadDate: (json['upload_date'] ?? '').toString(),
      createdAt: (json['created_at'] ?? '').toString(),
    );
  }
}

class CloudFeedbackItem {
  final int id;
  final String username;
  final String content;
  final String status;
  final String createdAt;

  const CloudFeedbackItem({
    required this.id,
    required this.username,
    required this.content,
    required this.status,
    required this.createdAt,
  });

  factory CloudFeedbackItem.fromJson(Map<String, dynamic> json) {
    return CloudFeedbackItem(
      id: json['id'] as int,
      username: (json['username'] ?? '').toString(),
      content: (json['content'] ?? '').toString(),
      status: (json['status'] ?? 'open').toString(),
      createdAt: (json['created_at'] ?? '').toString(),
    );
  }
}

class CloudAnimeAnalysisResult {
  final int? id;
  final String analysis;
  final String model;
  final String createdAt;
  final int? remainingToday;

  const CloudAnimeAnalysisResult({
    this.id,
    required this.analysis,
    required this.model,
    required this.createdAt,
    this.remainingToday,
  });

  factory CloudAnimeAnalysisResult.fromJson(Map<String, dynamic> json) {
    return CloudAnimeAnalysisResult(
      id: _asInt(json['id'] ?? json['record_id'] ?? json['server_record_id']),
      analysis: (json['analysis'] ?? json['content'] ?? json['text'] ?? '')
          .toString(),
      model: (json['model'] ?? 'deepseek-v4-flash').toString(),
      createdAt: (json['created_at'] ?? DateTime.now().toIso8601String())
          .toString(),
      remainingToday: _asInt(json['remaining_today']),
    );
  }

  static int? _asInt(dynamic value) {
    if (value == null) return null;
    if (value is int) return value;
    if (value is num) return value.toInt();
    return int.tryParse(value.toString());
  }
}

class CloudApiException implements Exception {
  final String message;

  const CloudApiException(this.message);

  @override
  String toString() => message;
}

class CloudAccountService {
  CloudAccountService._();

  static const maxBackupCount = 3;
  static const backupUploadCooldown = Duration(minutes: 5);

  static const _tokenKey = 'cloud_user_token';
  static const _usernameKey = 'cloud_username';
  static const _userIdKey = 'cloud_user_id';
  static const _lastBackupUploadAtKeyPrefix = 'cloud_last_backup_upload_at_';

  static Future<CloudSession?> loadSession() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString(_tokenKey);
    final username = prefs.getString(_usernameKey);
    final userId = prefs.getInt(_userIdKey);
    if (token == null || username == null || userId == null) return null;
    return CloudSession(userId: userId, username: username, token: token);
  }

  static Future<void> logout() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_tokenKey);
    await prefs.remove(_usernameKey);
    await prefs.remove(_userIdKey);
  }

  static Future<CloudSession> register({
    required String username,
    required String password,
    required String inviteCode,
  }) {
    return _auth('/api/auth/register', {
      'username': username,
      'password': password,
      'invite_code': inviteCode,
    });
  }

  static Future<CloudSession> login({
    required String username,
    required String password,
  }) {
    return _auth('/api/auth/login', {
      'username': username,
      'password': password,
    });
  }

  static Future<void> resetPassword({
    required String username,
    required String password,
    required String inviteCode,
  }) async {
    await _requestJson(
      method: 'POST',
      path: '/api/auth/reset-password',
      body: {
        'username': username,
        'password': password,
        'invite_code': inviteCode,
      },
    );
  }

  static Future<List<CloudBackupInfo>> listBackups(CloudSession session) async {
    final json = await _requestJson(
      method: 'GET',
      path: '/api/user/backups',
      session: session,
    );
    final data = json['data'] as List<dynamic>? ?? const [];
    final backups = data
        .map((e) => CloudBackupInfo.fromJson(e as Map<String, dynamic>))
        .toList();
    backups.sort(_compareBackupsNewestFirst);
    return backups.take(maxBackupCount).toList(growable: false);
  }

  static Future<CloudBackupInfo> uploadBackupFile({
    required CloudSession session,
    required File zipFile,
  }) async {
    final remaining = await getBackupUploadCooldownRemaining(session);
    if (remaining != null) {
      throw CloudApiException(
        '距离上次云存储未满 5 分钟，请 ${formatCooldown(remaining)} 后再试',
      );
    }

    final bytes = await zipFile.readAsBytes();
    final json = await _requestJson(
      method: 'POST',
      path: '/api/user/backups',
      session: session,
      body: {
        'file_name': path.basename(zipFile.path),
        'backup_base64': base64Encode(bytes),
      },
    );
    await _markBackupUploaded(session);
    return CloudBackupInfo.fromJson(json['data'] as Map<String, dynamic>);
  }

  static Future<Duration?> getBackupUploadCooldownRemaining(
    CloudSession session,
  ) async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_lastBackupUploadAtKey(session));
    if (raw == null || raw.isEmpty) return null;

    final lastUploadedAt = DateTime.tryParse(raw);
    if (lastUploadedAt == null) return null;

    final elapsedMs = DateTime.now().difference(lastUploadedAt).inMilliseconds;
    final remainingMs = backupUploadCooldown.inMilliseconds - elapsedMs;
    if (remainingMs <= 0) return null;

    final clampedRemainingMs = remainingMs > backupUploadCooldown.inMilliseconds
        ? backupUploadCooldown.inMilliseconds
        : remainingMs;
    return Duration(milliseconds: clampedRemainingMs);
  }

  static String formatCooldown(Duration duration) {
    final minutes = duration.inMinutes;
    final seconds = duration.inSeconds.remainder(60);
    if (minutes <= 0) return '$seconds 秒';
    if (seconds == 0) return '$minutes 分钟';
    return '$minutes 分 $seconds 秒';
  }

  static Future<File> downloadBackupFile({
    required CloudSession session,
    required CloudBackupInfo backup,
  }) async {
    final uri = ApiConfig.cloudUri('/api/user/backups/${backup.id}/download');
    if (uri == null) {
      throw const CloudApiException('尚未配置云端 API 地址');
    }

    final response = await http.get(
      uri,
      headers: {
        'Accept': 'application/zip, application/json',
        'Authorization': 'Bearer ${session.token}',
      },
    );
    if (response.statusCode < 200 || response.statusCode >= 300) {
      final responseJson = _decodeJson(utf8.decode(response.bodyBytes));
      final message = (responseJson['message'] ?? '下载云端备份失败').toString();
      throw CloudApiException(message);
    }

    final tempDir = await getTemporaryDirectory();
    final fileName = backup.fileName.endsWith('.zip')
        ? backup.fileName
        : 'AniMeow_cloud_backup_${backup.id}.zip';
    final file = File(path.join(tempDir.path, fileName));
    await file.writeAsBytes(response.bodyBytes, flush: true);
    return file;
  }

  static Future<List<CloudFeedbackItem>> listFeedback(
    CloudSession session,
  ) async {
    final json = await _requestJson(
      method: 'GET',
      path: '/api/feedback',
      session: session,
    );
    final data = json['data'] as List<dynamic>? ?? const [];
    return data
        .map((e) => CloudFeedbackItem.fromJson(e as Map<String, dynamic>))
        .toList(growable: false);
  }

  static Future<CloudFeedbackItem> submitFeedback({
    required CloudSession session,
    required String content,
  }) async {
    final json = await _requestJson(
      method: 'POST',
      path: '/api/feedback',
      session: session,
      body: {'content': content},
    );
    return CloudFeedbackItem.fromJson(json['data'] as Map<String, dynamic>);
  }

  static Future<CloudAnimeAnalysisResult> analyzeAnimeStats({
    required CloudSession session,
    required Map<String, dynamic> stats,
    String? clientVersion,
  }) async {
    final json = await _requestJson(
      method: 'POST',
      path: '/api/user/anime-analysis',
      session: session,
      body: {
        'model': 'deepseek-v4-flash',
        if (clientVersion != null && clientVersion.isNotEmpty)
          'client_version': clientVersion,
        'stats': stats,
      },
    );
    final data = json['data'];
    if (data is Map<String, dynamic>) {
      return CloudAnimeAnalysisResult.fromJson(data);
    }
    return CloudAnimeAnalysisResult.fromJson(json);
  }

  static Future<CloudSession> _auth(
    String path,
    Map<String, dynamic> body,
  ) async {
    final json = await _requestJson(method: 'POST', path: path, body: body);
    final data = json['data'] as Map<String, dynamic>;
    final user = data['user'] as Map<String, dynamic>;
    final session = CloudSession(
      userId: (user['id'] as num).toInt(),
      username: (user['username'] ?? '').toString(),
      token: (data['token'] ?? '').toString(),
    );
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_tokenKey, session.token);
    await prefs.setString(_usernameKey, session.username);
    await prefs.setInt(_userIdKey, session.userId);
    return session;
  }

  static Future<Map<String, dynamic>> _requestJson({
    required String method,
    required String path,
    CloudSession? session,
    Map<String, dynamic>? body,
  }) async {
    final uri = ApiConfig.cloudUri(path);
    if (uri == null) {
      throw const CloudApiException('尚未配置云端 API 地址');
    }

    final headers = <String, String>{
      'Accept': 'application/json',
      if (body != null) 'Content-Type': 'application/json',
      if (session != null) 'Authorization': 'Bearer ${session.token}',
    };

    final encodedBody = body == null ? null : jsonEncode(body);
    final response = switch (method) {
      'GET' => await http.get(uri, headers: headers),
      'POST' => await http.post(uri, headers: headers, body: encodedBody),
      _ => throw CloudApiException('不支持的请求方法：$method'),
    };

    final responseJson = _decodeJson(response.body);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      final message = (responseJson['message'] ?? '云端请求失败').toString();
      throw CloudApiException(message);
    }
    if (responseJson['status'] != 'success') {
      final message = (responseJson['message'] ?? '云端请求失败').toString();
      throw CloudApiException(message);
    }
    return responseJson;
  }

  static Map<String, dynamic> _decodeJson(String body) {
    if (body.isEmpty) return const {};
    final decoded = jsonDecode(body);
    if (decoded is Map<String, dynamic>) return decoded;
    return const {};
  }

  static int _compareBackupsNewestFirst(CloudBackupInfo a, CloudBackupInfo b) {
    final aTime = _backupTime(a);
    final bTime = _backupTime(b);
    if (aTime != null && bTime != null) return bTime.compareTo(aTime);
    if (aTime != null) return -1;
    if (bTime != null) return 1;
    return b.id.compareTo(a.id);
  }

  static DateTime? _backupTime(CloudBackupInfo backup) {
    return DateTime.tryParse(backup.createdAt) ??
        DateTime.tryParse(backup.uploadDate);
  }

  static Future<void> _markBackupUploaded(CloudSession session) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(
      _lastBackupUploadAtKey(session),
      DateTime.now().toIso8601String(),
    );
  }

  static String _lastBackupUploadAtKey(CloudSession session) {
    return '$_lastBackupUploadAtKeyPrefix${session.userId}';
  }
}

File: android_app/app/src/main/java/com/animeow/app/data/cloud/CloudAccountService.kt
package com.animeow.app.data.cloud

import android.content.Context
import android.net.Uri
import android.util.Base64
import androidx.core.content.FileProvider
import com.animeow.app.BuildConfig
import com.animeow.app.data.backup.NativeBackupFormatException
import com.animeow.app.data.backup.NativeBackupService
import com.animeow.app.data.backup.NativeBackupSummary
import com.animeow.app.data.backup.NativeRestorePoint
import com.animeow.app.data.backup.NativeRestoreSelection
import com.animeow.app.data.backup.NativeRestoreStrategy
import com.animeow.app.data.backup.PreparedNativeRestore
import com.animeow.app.data.importer.LegacyBackupImporter
import com.animeow.app.data.importer.LegacyImportSummary
import com.animeow.app.data.importer.PreparedLegacyImport
import com.animeow.app.data.remote.JsonHttpClient
import com.animeow.app.data.remote.RemoteHttpException
import java.io.File
import java.time.Instant
import org.json.JSONArray
import org.json.JSONObject

data class CloudBackupInfo(
    val id: Long,
    val fileName: String,
    val payloadSize: Long,
    val uploadDate: String,
    val createdAt: String,
    val contentHash: String? = null,
    val deviceId: String? = null,
    val clientModifiedAt: String? = null,
)

data class CloudFeedbackItem(
    val id: Long,
    val username: String,
    val content: String,
    val status: String,
    val reply: String?,
    val replyAt: String?,
    val isDeveloper: Boolean,
    val createdAt: String,
)

data class CloudFeedbackPool(
    val items: List<CloudFeedbackItem>,
    val viewerIsAdmin: Boolean,
    val viewerIsDeveloper: Boolean,
)

data class CloudAnimeAnalysisResult(
    val id: Long,
    val model: String,
    val analysis: String,
    val createdAt: String,
)

enum class CloudBackupFormat {
    NATIVE,
    LEGACY,
}

data class CloudRestorePreview(
    val animeCount: Int,
    val characterCount: Int,
    val tagCount: Int,
    val coverCount: Int,
)

class PreparedCloudRestore internal constructor(
    val backup: CloudBackupInfo,
    val format: CloudBackupFormat,
    val preview: CloudRestorePreview,
    internal val nativePrepared: PreparedNativeRestore? = null,
    internal val legacyPrepared: PreparedLegacyImport? = null,
    internal val legacySummary: LegacyImportSummary? = null,
    internal var restorePoint: NativeRestorePoint? = null,
) {
    internal val nativeSummary: NativeBackupSummary?
        get() = nativePrepared?.summary
}

class PreparedCloudUpload internal constructor(
    internal val file: File,
    val summary: NativeBackupSummary,
)

data class CloudRestoreResult(
    val summary: NativeBackupSummary,
    val restorePoint: NativeRestorePoint,
)

internal class CloudAccountService(
    context: Context,
    private val nativeBackupService: NativeBackupService,
    private val legacyBackupImporter: LegacyBackupImporter,
    private val client: JsonHttpClient = JsonHttpClient(),
) {
    private val appContext = context.applicationContext
    private val sessionStore = SecureCloudSessionStore(appContext)
    private val deviceRegistrationId = DeviceRegistrationId.from(appContext)

    val isConfigured: Boolean get() = BuildConfig.CLOUD_API_BASE.isNotBlank()

    fun loadSession(): CloudSession? = sessionStore.load()

    fun logout() = sessionStore.clear()

    suspend fun register(username: String, password: String): CloudSession = authenticate(
        path = "/api/auth/register",
        body = JSONObject()
            .put("username", username.trim())
            .put("password", password)
            .put("device_registration_id", deviceRegistrationId),
    )

    suspend fun login(username: String, password: String): CloudSession = authenticate(
        path = "/api/auth/login",
        body = JSONObject().put("username", username.trim()).put("password", password),
    )

    suspend fun resetPassword(username: String, password: String, inviteCode: String) {
        val body = JSONObject()
            .put("username", username.trim())
            .put("password", password)
            .put("device_registration_id", deviceRegistrationId)
        inviteCode.trim().takeIf(String::isNotEmpty)?.let {
            body.put("invite_code", it.uppercase(java.util.Locale.ROOT))
        }
        requestJson(
            method = "POST",
            path = "/api/auth/reset-password",
            body = body,
        )
    }

    suspend fun listBackups(session: CloudSession): List<CloudBackupInfo> {
        val root = requestJson("GET", "/api/user/backups", session)
        return root.optJSONArray("data").toCloudBackups().sortedByDescending(CloudBackupInfo::id)
    }

    suspend fun prepareUpload(): PreparedCloudUpload {
        val directory = File(appContext.cacheDir, "shared/cloud_upload").apply { mkdirs() }
        val file = File(directory, "AniMeow_${System.currentTimeMillis()}.animeow.zip")
        try {
            val summary = nativeBackupService.export(file)
            require(file.length() in 1..MAX_CLOUD_BACKUP_BYTES) {
                "云端备份需小于 ${MAX_CLOUD_BACKUP_BYTES / 1024 / 1024} MB"
            }
            return PreparedCloudUpload(file, summary)
        } catch (error: Throwable) {
            file.delete()
            throw error
        }
    }

    suspend fun uploadPrepared(
        session: CloudSession,
        prepared: PreparedCloudUpload,
        deviceId: String? = null,
    ): CloudBackupInfo {
        try {
            // Read file once, compute SHA-256 of the actual ZIP bytes being
            // uploaded.  The server independently computes sha256(payload)
            // on the decoded bytes, so the hash must be over the file content,
            // not over the logical backup fingerprint.
            val fileBytes = prepared.file.readBytes()
            val contentHash = java.security.MessageDigest.getInstance("SHA-256")
                .digest(fileBytes)
                .joinToString("") { "%02x".format(it) }
            val body = JSONObject()
                .put("file_name", prepared.file.name)
                .put("backup_base64", Base64.encodeToString(fileBytes, Base64.NO_WRAP))
                .put("content_hash", contentHash)
                .put("client_modified_at", prepared.summary.createdAt)
            deviceId?.takeIf(String::isNotBlank)?.let { body.put("device_id", it) }
            val root = requestJson("POST", "/api/user/backups", session, body)
            return root.getJSONObject("data").toCloudBackup()
        } finally {
            prepared.file.delete()
        }
    }

    suspend fun uploadBackup(session: CloudSession, deviceId: String? = null): CloudBackupInfo =
        uploadPrepared(session, prepareUpload(), deviceId)

    fun discard(prepared: PreparedCloudUpload) {
        prepared.file.delete()
    }

    suspend fun prepareRestore(session: CloudSession, backup: CloudBackupInfo): PreparedCloudRestore {
        val bytes = try {
            client.getBytes(
                url = baseUrl() + "/api/user/backups/${backup.id}/download",
                headers = mapOf("Authorization" to "Bearer ${session.token}", "Accept" to "application/zip"),
            )
        } catch (exception: RemoteHttpException) {
            throwRemoteFailure(exception, authenticated = true)
        }
        require(bytes.size.toLong() in 1..MAX_CLOUD_BACKUP_BYTES) { "云端备份文件大小异常" }
        val directory = File(appContext.cacheDir, "shared/cloud_restore").apply { mkdirs() }
        val file = File(directory, "restore_${backup.id}_${System.currentTimeMillis()}.zip")
        return try {
            file.writeBytes(bytes)
            val uri = fileProviderUri(file)
            try {
                val native = nativeBackupService.prepareRestore(uri)
                PreparedCloudRestore(
                    backup = backup,
                    format = CloudBackupFormat.NATIVE,
                    preview = CloudRestorePreview(
                        animeCount = native.summary.animeCount,
                        characterCount = native.summary.characterCount,
                        tagCount = native.summary.tagCount,
                        coverCount = native.summary.coverCount,
                    ),
                    nativePrepared = native,
                )
            } catch (nativeFormatError: NativeBackupFormatException) {
                val legacy = try {
                    legacyBackupImporter.prepare(uri)
                } catch (legacyError: Throwable) {
                    nativeFormatError.addSuppressed(legacyError)
                    throw nativeFormatError
                }
                PreparedCloudRestore(
                    backup = backup,
                    format = CloudBackupFormat.LEGACY,
                    preview = CloudRestorePreview(
                        animeCount = legacy.summary.animeCount,
                        characterCount = legacy.summary.characterCount,
                        tagCount = legacy.summary.tagCount,
                        coverCount = legacy.summary.coverCount,
                    ),
                    legacyPrepared = legacy,
                    legacySummary = legacy.summary,
                )
            }
        } finally {
            file.delete()
        }
    }

    suspend fun restore(prepared: PreparedCloudRestore): CloudRestoreResult {
        return restore(prepared, NativeRestoreSelection())
    }

    suspend fun restore(
        prepared: PreparedCloudRestore,
        selection: NativeRestoreSelection,
    ): CloudRestoreResult {
        return when (prepared.format) {
            CloudBackupFormat.NATIVE -> {
                val native = requireNotNull(prepared.nativePrepared) { "原生云备份预检状态无效" }
                val restorePoint = prepared.restorePoint
                    ?: nativeBackupService.createRestorePoint("before_cloud_${prepared.backup.id}")
                        .also { prepared.restorePoint = it }
                CloudRestoreResult(
                    summary = nativeBackupService.restore(native, selection),
                    restorePoint = restorePoint,
                )
            }
            CloudBackupFormat.LEGACY -> {
                require(selection.libraryAndCharacters) {
                    "旧版云备份只包含资料库数据，请选择恢复资料库与封面"
                }
                val legacy = requireNotNull(prepared.legacyPrepared) { "旧版云备份预检状态无效" }
                val restorePoint = prepared.restorePoint
                    ?: nativeBackupService.createRestorePoint("before_cloud_legacy_${prepared.backup.id}")
                        .also {
                            prepared.restorePoint = it
                            legacy.restorePointCreated = true
                        }
                legacyBackupImporter.importPrepared(legacy)
                CloudRestoreResult(
                    summary = nativeBackupService.summarizeCurrent(),
                    restorePoint = restorePoint,
                )
            }
        }
    }

    suspend fun merge(
        prepared: PreparedCloudRestore,
        selection: NativeRestoreSelection,
    ): CloudRestoreResult {
        require(prepared.format == CloudBackupFormat.NATIVE) {
            "旧版云备份只能完整迁移，不能与 2.0 数据逐项合并"
        }
        val native = requireNotNull(prepared.nativePrepared) { "原生云备份预检状态无效" }
        val restorePoint = prepared.restorePoint
            ?: nativeBackupService.createRestorePoint("before_cloud_merge_${prepared.backup.id}")
                .also { prepared.restorePoint = it }
        return CloudRestoreResult(
            summary = nativeBackupService.restore(
                prepared = native,
                selection = selection,
                strategy = NativeRestoreStrategy.MERGE_KEEP_LOCAL,
            ),
            restorePoint = restorePoint,
        )
    }

    fun discard(prepared: PreparedCloudRestore) {
        prepared.nativePrepared?.let(nativeBackupService::discard)
        prepared.legacyPrepared?.let(legacyBackupImporter::discard)
    }

    suspend fun deleteBackup(session: CloudSession, backup: CloudBackupInfo) {
        requestJson("DELETE", "/api/user/backups/${backup.id}", session)
    }

    fun listRestorePoints(): List<NativeRestorePoint> = nativeBackupService.listRestorePoints()

    suspend fun prepareRestorePoint(point: NativeRestorePoint): PreparedNativeRestore =
        nativeBackupService.prepareRestore(point.file)

    suspend fun restorePoint(prepared: PreparedNativeRestore): NativeBackupSummary =
        nativeBackupService.restore(prepared)

    fun discardRestorePoint(prepared: PreparedNativeRestore) = nativeBackupService.discard(prepared)

    suspend fun listFeedback(session: CloudSession? = null): CloudFeedbackPool {
        val root = requestJson("GET", "/api/feedback", session)
        val data = root.optJSONArray("data") ?: JSONArray()
        val viewer = root.optJSONObject("viewer")
        val items = buildList {
            for (index in 0 until data.length()) {
                data.optJSONObject(index)?.let { json ->
                    add(
                        CloudFeedbackItem(
                            id = json.optLong("id"),
                            username = json.optString("username"),
                            content = json.optString("content"),
                            status = json.optString("status", "open"),
                            reply = if (json.isNull("reply")) null else json.optString("reply").takeIf(String::isNotBlank),
                            replyAt = if (json.isNull("reply_at")) null else json.optString("reply_at").takeIf(String::isNotBlank),
                            isDeveloper = json.optBoolean("is_developer"),
                            createdAt = json.optString("created_at"),
                        ),
                    )
                }
            }
        }
        return CloudFeedbackPool(
            items = items,
            viewerIsAdmin = viewer?.optBoolean("is_admin") == true,
            viewerIsDeveloper = viewer?.optBoolean("is_developer") == true,
        )
    }

    suspend fun submitFeedback(session: CloudSession, content: String): CloudFeedbackItem {
        val clean = content.trim()
        require(clean.length in 5..1000) { "反馈内容需为 5-1000 字" }
        val json = requestJson("POST", "/api/feedback", session, JSONObject().put("content", clean))
            .getJSONObject("data")
        return CloudFeedbackItem(
            id = json.optLong("id"),
            username = json.optString("username", session.username),
            content = json.optString("content", clean),
            status = json.optString("status", "open"),
            reply = if (json.isNull("reply")) null else json.optString("reply").takeIf(String::isNotBlank),
            replyAt = if (json.isNull("reply_at")) null else json.optString("reply_at").takeIf(String::isNotBlank),
            isDeveloper = json.optBoolean("is_developer"),
            createdAt = json.optString("created_at", Instant.now().toString()),
        )
    }

    suspend fun setFeedbackStatus(session: CloudSession, feedbackId: Long, status: String) {
        requestJson(
            "POST",
            "/api/feedback/$feedbackId/status",
            session,
            JSONObject().put("status", status),
        )
    }

    suspend fun replyFeedback(session: CloudSession, feedbackId: Long, reply: String) {
        requestJson(
            "POST",
            "/api/feedback/$feedbackId/reply",
            session,
            JSONObject().put("reply", reply.trim()),
        )
    }

    suspend fun analyzeAnimeStats(
        session: CloudSession,
        stats: JSONObject,
        clientVersion: String,
    ): CloudAnimeAnalysisResult {
        val data = requestJson(
            method = "POST",
            path = "/api/user/anime-analysis",
            session = session,
            body = JSONObject()
                .put("stats", stats)
                .put("client_version", clientVersion.take(32)),
        ).getJSONObject("data")
        return CloudAnimeAnalysisResult(
            id = data.optLong("id"),
            model = data.optString("model", "server"),
            analysis = data.optString("analysis").ifBlank { error("云端模型没有返回分析内容") },
            createdAt = data.optString("created_at", Instant.now().toString()),
        )
    }

    private suspend fun authenticate(path: String, body: JSONObject): CloudSession {
        require(body.optString("username").matches(Regex("[A-Za-z0-9_]{3,24}"))) {
            "用户名需为 3-24 位字母、数字或下划线"
        }
        require(body.optString("password").length in 6..72) { "密码需为 6-72 位" }
        val data = requestJson("POST", path, body = body).getJSONObject("data")
        val user = data.getJSONObject("user")
        val session = CloudSession(
            userId = user.getLong("id"),
            username = user.getString("username"),
            token = data.getString("token"),
        )
        sessionStore.save(session)
        return session
    }

    private suspend fun requestJson(
        method: String,
        path: String,
        session: CloudSession? = null,
        body: JSONObject? = null,
    ): JSONObject {
        val headers = buildMap {
            put("Accept", "application/json")
            session?.let { put("Authorization", "Bearer ${it.token}") }
        }
        val response = try {
            when (method) {
                "GET" -> client.get(baseUrl() + path, headers)
                "POST" -> client.post(baseUrl() + path, body?.toString().orEmpty(), headers)
                "PUT" -> client.put(baseUrl() + path, body?.toString().orEmpty(), headers)
                "DELETE" -> client.delete(baseUrl() + path, headers)
                else -> error("不支持的云端请求方法")
            }
        } catch (exception: RemoteHttpException) {
            throwRemoteFailure(exception, authenticated = session != null)
        }
        val root = JSONObject(response)
        if (root.optString("status") != "success") {
            error(safeCloudMessage(root.optString("message")) ?: "云端请求失败，请稍后重试")
        }
        return root
    }

    private fun baseUrl(): String = BuildConfig.CLOUD_API_BASE.trim().trimEnd('/')
        .takeIf(String::isNotEmpty)
        ?: error("当前构建尚未配置云端 API 地址")

    private fun throwRemoteFailure(exception: RemoteHttpException, authenticated: Boolean): Nothing {
        val failure = cloudRemoteFailure(
            statusCode = exception.statusCode,
            responseBody = exception.responseBody,
            authenticated = authenticated,
        )
        if (failure is CloudSessionExpiredException) sessionStore.clear()
        throw failure
    }

    private fun fileProviderUri(file: File): Uri = FileProvider.getUriForFile(
        appContext,
        "${appContext.packageName}.fileprovider",
        file,
    )

    private fun JSONObject.toCloudBackup(): CloudBackupInfo = CloudBackupInfo(
        id = getLong("id"),
        fileName = optString("file_name", "AniMeow_cloud.zip"),
        payloadSize = optLong("payload_size"),
        uploadDate = optString("upload_date"),
        createdAt = optString("created_at"),
        contentHash = optString("content_hash").takeIf(String::isNotBlank),
        deviceId = optString("device_id").takeIf(String::isNotBlank),
        clientModifiedAt = optString("client_modified_at").takeIf(String::isNotBlank),
    )

    private fun JSONArray?.toCloudBackups(): List<CloudBackupInfo> {
        val array = this ?: JSONArray()
        return buildList {
            for (index in 0 until array.length()) {
                array.optJSONObject(index)?.let { add(it.toCloudBackup()) }
            }
        }
    }

    private companion object {
        const val MAX_CLOUD_BACKUP_BYTES = 20L * 1024 * 1024
    }
}
